resource/kind.go

```diff
@@ -22,334 +18,162 @@ const (
-// See TypedKind and UntypedKind as examples of KindWriter implementations.
-type KindWriter interface {
-	// Write consumes an Object-implementation of an instance of the Kind, and writes marshaled wire-format bytes
-	// to the provided io.Writer. It MAY return an error if the provided Object is not of the expected underlying type(s),
-	// and MUST return an error if the object cannot be marshaled into bytes.
-	Write(obj Object, out io.Writer, encoding KindEncoding) error
-}
-
-// KindReadWriter is an interface that combines KindReader and KindWriter
-type KindReadWriter interface {
-	KindReader
-	KindWriter
-}
-
-// Kind is an interface representing a kubernetes-compatible Kind, which contains information about the Kind's
-// Group, Version, Kind, Plural, and Scope, as well as allowing for creation of a zero-value version of the Kind as an
-// Object (which may contain default values for fields as decided by the implementation).
-// Additionally, Kind contains the logic for reading wire bytes into an Object instance of the Kind,
-// or serializing an Object instance of the Kind into wire bytes.
-// In this sense, Kind combines schema information and a serializer into one interface used to interact with the kind.
-type Kind interface {
-	KindReadWriter
+// SchemaWithCodecs is an interface which *Kind implements, and may be used in place of Kind.
+// TODO: On the fence about whether stores/clients should accept Kind or SchemaWithCodecs
+type SchemaWithCodecs interface {
 	Schema
+	Codec(KindEncoding) Codec
 }
 
-// NewKindGroup returns a new KindGroup with the provided group and version
-func NewKindGroup(group, version string) *KindGroup {
-	return &KindGroup{
-		group:   group,
-		version: version,
-		kinds:   make([]Kind, 0),
-	}
-}
-
-// KindGroup is a set of Kind objects which share the same Group and Version.
-type KindGroup struct {
-	group   string
-	version string
-	kinds   []Kind
-}
-
-// Group returns the group value shared by all kinds in the KindGroup
-func (k *KindGroup) Group() string {
-	return k.group
-}
-
-// Version returns the version value shared by all kinds in the KindGroup
-func (k *KindGroup) Version() string {
-	return k.version
+// Codec is an interface which describes any object which can read and write Object implementations to/from bytes.
+// A codec is often specific to an encoding of the bytes in the reader/writer, and may also be specific to
+// Object implementations.
+type Codec interface {
+	Read(in io.Reader, into Object) error
+	Write(out io.Writer, obj Object) error
 }
 
-// Kinds returns the list of kinds registered with the group
-func (k *KindGroup) Kinds() []Kind {
-	return k.kinds
+// Kind is a struct which encapsulates Schema information and Codecs for reading/writing Objects which are instances
+// of the contained Schema. It implements Schema using the Schema field.
+type Kind struct {
+	Schema Schema
+	Codecs map[KindEncoding]Codec
 }
```

Comment:
NIT: seeing as we're just re-implementing `Schema` methods in the Kind, we might as well just embed it. As much as I try to avoid embedding types I think it's one of the few cases where it makes sense.
```suggestion
type Kind struct {
	Schema
	Codecs map[KindEncoding]Codec
}
```